pkg/services/preference/prefimpl/pref.go

```diff
@@ -0,0 +1,203 @@
+	preference, err := s.store.Get(ctx, &pref.Preference{
+		OrgID:  cmd.OrgID,
+		UserID: cmd.UserID,
+		TeamID: cmd.TeamID,
+	})
+	if err != nil {
+		if errors.Is(err, pref.ErrPrefNotFound) {
+			preference := &pref.Preference{
+				UserID:          cmd.UserID,
+				OrgID:           cmd.OrgID,
+				TeamID:          cmd.TeamID,
+				HomeDashboardID: cmd.HomeDashboardID,
+				Timezone:        cmd.Timezone,
+				WeekStart:       cmd.WeekStart,
+				Theme:           cmd.Theme,
+				Created:         time.Now(),
+				Updated:         time.Now(),
+			}
+			_, err = s.store.Insert(ctx, preference)
+			if err != nil {
+				return err
+			}
+		}
+		return err
+	}
+	preference.Timezone = cmd.Timezone
+	preference.WeekStart = cmd.WeekStart
+	preference.Theme = cmd.Theme
+	preference.Updated = time.Now()
+	preference.Version += 1
+	preference.JSONData = &pref.PreferenceJSONData{}
+
+	if cmd.Navbar != nil {
+		preference.JSONData.Navbar = *cmd.Navbar
+	}
+	if cmd.QueryHistory != nil {
+		preference.JSONData.QueryHistory = *cmd.QueryHistory
+	}
+	return s.store.Update(ctx, preference)
+}
+
+func (s *Service) Patch(ctx context.Context, cmd *pref.PatchPreferenceCommand) error {
+	var exists bool
+	preference, err := s.store.Get(ctx, &pref.Preference{
+		OrgID:  cmd.OrgID,
+		UserID: cmd.UserID,
+		TeamID: cmd.TeamID,
+	})
+	if err != nil && !errors.Is(err, pref.ErrPrefNotFound) {
+		return err
+	}
+
+	if errors.Is(err, pref.ErrPrefNotFound) {
+		preference = &pref.Preference{
+			UserID:   cmd.UserID,
+			OrgID:    cmd.OrgID,
+			TeamID:   cmd.TeamID,
+			Created:  time.Now(),
+			JSONData: &pref.PreferenceJSONData{},
+		}
+	} else {
+		exists = true
+	}
+
+	if cmd.Navbar != nil {
+		if preference.JSONData == nil {
+			preference.JSONData = &pref.PreferenceJSONData{}
+		}
+		if cmd.Navbar.SavedItems != nil {
+			preference.JSONData.Navbar.SavedItems = cmd.Navbar.SavedItems
+		}
+	}
+
+	if cmd.QueryHistory != nil {
+		if preference.JSONData == nil {
+			preference.JSONData = &pref.PreferenceJSONData{}
+		}
+		if cmd.QueryHistory.HomeTab != "" {
+			preference.JSONData.QueryHistory.HomeTab = cmd.QueryHistory.HomeTab
+		}
+	}
+
+	if cmd.HomeDashboardID != nil {
+		preference.HomeDashboardID = *cmd.HomeDashboardID
+	}
+
+	if cmd.Timezone != nil {
+		preference.Timezone = *cmd.Timezone
+	}
+
+	if cmd.WeekStart != nil {
+		preference.WeekStart = *cmd.WeekStart
+	}
+
+	if cmd.Theme != nil {
+		preference.Theme = *cmd.Theme
+	}
+
+	preference.Updated = time.Now()
+	preference.Version += 1
+
+	// Wrap this in an if statement to maintain backwards compatibility
+	if cmd.Navbar != nil {
+		if preference.JSONData == nil {
+			preference.JSONData = &pref.PreferenceJSONData{}
+		}
+		if cmd.Navbar.SavedItems != nil {
+			preference.JSONData.Navbar.SavedItems = cmd.Navbar.SavedItems
+		}
+	}
+
+	if exists {
+		err = s.store.Update(ctx, preference)
+	} else {
+		_, err = s.store.Insert(ctx, preference)
```

Comment:
Is `Insert` something a `Patch` operation should be able to do?

Reply:
I kept the naming and the functionality from sqlstore

Reply:
I agree that at the Store level we should try to keep names closer to CRUD, since most of the possible Store implementations use that, while Patch is very unusual (i.e. S3 doesn't allow you to patch an object). Maybe even narrow it down to Put/Get/Delete/List. But I'm fine with the current naming, too.

Reply:
this is the endpoint PATCH /api/user/preferences, maybe we can rename on the PR where we implement the service in the api